docs/source/sparse.rst

```diff
@@ -7,51 +7,162 @@
+
+Why and when to use sparsity
+++++++++++++++++++++++++++++
+
+By default PyTorch stores :class:`torch.Tensor` stores elements contiguously
+physical memory. This leads to efficient implementations of various array
+processing algorithms that require fast access to elements.
+
+Now, some users might decide to represent data such as graph adjacency
+matrices, pruned weights or points clouds by Tensors whose *elements are
+mostly zero valued*. We recognize these are important applications and aim
+to provide performance optimizations for these use cases via sparse storage formats.
+
+Various sparse storage formats such as COO, CSR/CSC, LIL, etc.
+have been developed over the years. In a nutshell, they all compress data by
+tracking the positions of non-zero valued elements and perhaps some,
+but *most importantly* not all, zero valued elements. We call the uncompressed
+values *specified*. They *may* be zero. In contrast *unspecified* elements
+*must* be zero.
+
+By compressing repeat zeros sparse storage formats aim to save memory
+and computational resources on various CPUs and GPUs. Especially for high
+degrees of sparsity or highly structured sparsity this can have significant
+performance implications. As such sparse storage formats can be seen as a
+performance optimization.
+
+Like many other performance optimization sparse storage formats are not
+always advantageous. Details such as the distribution of zero valued
+elements, the level of sparsity and the user's device matter.
+
+We aim to make it easy to try these different compression techniques,
+but we do not yet provide a programmatic means of determining the best one
+for your particular application. Of course you are always welcome to
+ask for help in our discussion forum https://dev-discuss.pytorch.org/.
+
+Functionality overview
+++++++++++++++++++++++
+
+We want it to be straightforward to construct a sparse Tensor from a
+given dense Tensor by providing conversion routines for each layout.
+
+In the next example we convert a 2D Tensor with default strided layout
+to a 2D Tensor backed by the COO memory layout. Only values and indices
+of non-zero elements are stored in this case.
+
+    >>> a = torch.tensor([[0, 2.], [3, 0]])
+    >>> a.to_sparse()
+    tensor(indices=tensor([[0, 1],
+                           [1, 0]]),
+           values=tensor([2., 3.]),
+           size=(2, 2), nnz=2, layout=torch.sparse_coo)
+
+PyTorch currently supports :ref:`COO<sparse-coo-docs>`, :ref:`CSR<sparse-csr-docs>`,
+:ref:`CSC<sparse-csc-docs>`, :ref:`BSR<sparse-bsr-docs>`, and :ref:`BSC<sparse-bsc-docs>`.
+Please see the references for more details.
+
+Note that we provide slight generalizations of these formats.
+
+Batching: Devices such as GPUs require batching for optimal performance and
+thus we support batch dimensions.
+
+In this example we construct a 3D CSR Tensor from a 3D strided Tensor.
+
+    >>> t = torch.tensor([[[1., 0], [2., 3.]], [[4., 0], [5., 6.]]])
```

Comment:
minor nit, but It might be more interesting to show a tensor whose sparse compressed form looks significantly smaller. That is, the strided tensor is (in this case) mostly column 0s. Same comment for the prior example.